### PR TITLE
Refactor the OrderByKeyToClassConstRector to use the new enum only in `Criteria::orderBy` method calls

### PR DESCRIPTION
**Pull request by julienfastre**
Since version 2.2.0 of `doctrine/collection`, the constant `Doctrine\Common\Collections\Criteria` is deprecated in favor of new enum `\Doctrine\Common\Collections\Order`.

But this enum should only be used in invocations of `Doctrine\Common\Collection\Criteria::orderBy` method call. Other calls (in attributes, or for the same method name on object instantiating another class) should not be changed: the string `'ASC'` and `'DESC'` should be kept elsewhere.

This PR refactor the rule which, previously, replaced all the `'ASC'` and `'DESC'` by the `Criteria::ASC|DESC`. 

After this MR:

- all invocation of class constant `Criteria::ASC` and `Criteria::DESC` are replaced by, respectively, the string `"ASC"` ans `"DESC"`;
- in the method invocation `Criteria::orderBy`, the usage of the `"ASC"` (lowercase or uppercase), `"DESC"` (lowercase or uppercase), `Criteria::ASC`, `Criteria::DESC` is replaced by the new enum `Order::Ascending`, `Order::Descending`;
- the rule is added to the new ruleset for Collection 2.2;

The previous rule, which changed the usage of the string `"ASC"` (lowercase or uppercase) or `"DESC"` (lowercase or uppercase) by the usage of `Criteria::ASC` and `Criteria::DESC` is, obviously, removed.

## Comments and documentation supporting this change

- https://github.com/doctrine/collections/pull/389;
- https://github.com/doctrine/orm/issues/11313#issuecomment-2214570323

Closes https://github.com/rectorphp/rector-doctrine/issues/325

**EDIT** list the changes in a more understandable (?) way.

### CODE REVIEW

**Comment by julienfastre** [comment]
Still some TODO in this PR:

- [ ]  this rule should not apply on PHP versions earlier that 8.1.0 and doctrine collection greater that 2.2.0
- [ ]  find a way to ensure that the 'orderBy' method does apply on a \Doctrine\Common\Collections\Criteria instance
- [ ]  find a way to identify the full FQDN for Criteria

File: rules/Collection22/Rector/CriteriaOrderingConstantsDeprecationRector.php [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+        if (! $this->isName($node->name, 'orderBy')) {
+            return null;
+        }
+
+        $isSuperTypeOfCriteria = $this->criteriaObjectType->isSuperTypeOf($this->nodeTypeResolver->getType($node->var));
+
+        if ($isSuperTypeOfCriteria->no() || $isSuperTypeOfCriteria->maybe()) {
+            return null;
+        }
+
```
Comment (samsonasik) [inline rules/Collection22/Rector/CriteriaOrderingConstantsDeprecationRector.php:139]: check first class callable first

```php
if ($node->isFirstClassCallable()) {
    return null;
}
```

File: rules/Collection22/Rector/CriteriaOrderingConstantsDeprecationRector.php [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+        if ($node->isFirstClassCallable()) {
+            return null;
+        }
+
+        $isSuperTypeOfCriteria = $this->criteriaObjectType->isSuperTypeOf($this->nodeTypeResolver->getType($node->var));
+
+        if ($isSuperTypeOfCriteria->no() || $isSuperTypeOfCriteria->maybe()) {
```
Comment (samsonasik) [inline rules/Collection22/Rector/CriteriaOrderingConstantsDeprecationRector.php:140]: ! yes() should cover both no() and maybe()

File: rules/Collection22/Rector/CriteriaOrderingConstantsDeprecationRector.php [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+
+        return null;
+    }
+
+    /**
+     * @param 'ASC'|'DESC' $direction
+     * @return ArrayItem
+     */
+    private function buildArrayItem(string $direction, Node\Expr|null $key): Node\Expr\ArrayItem
+    {
```
Comment (samsonasik) [inline rules/Collection22/Rector/CriteriaOrderingConstantsDeprecationRector.php:208]: $direction need to be transformed to strtoupper first to ensure it always uppercase to compare

Comment (julienfastre) [reply rules/Collection22/Rector/CriteriaOrderingConstantsDeprecationRector.php:206]: Actually, this is done on lines [169](https://github.com/rectorphp/rector-doctrine/pull/336/files#diff-8248dc7250edd8eee798927000b8dd81be8511a938eb9a8a2ee6dc44e7f31317R169) and [181](https://github.com/rectorphp/rector-doctrine/pull/336/files#diff-8248dc7250edd8eee798927000b8dd81be8511a938eb9a8a2ee6dc44e7f31317R181). 

But this is maybe not the most readable way for doing this...

**Review by samsonasik — APPROVED** [review]
Ok, let's give it a try 👍

**Comment by samsonasik** [comment]
Thank you @julienfastre

File: rules/Collection22/Rector/CriteriaOrderingConstantsDeprecationRector.php [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+    }
+
+    public function getRuleDefinition(): RuleDefinition
+    {
+        return new RuleDefinition(
+            'Replace ASC/DESC with enum Ordering in Criteria::orderBy method call, and remove usage of Criteria::ASC and Criteria::DESC constants elsewhere',
+            [
+                new CodeSample(
+                    <<<'CODE_SAMPLE'
+                    <?php
```
Comment (samsonasik) [inline rules/Collection22/Rector/CriteriaOrderingConstantsDeprecationRector.php:35]: php open tag seems cause use statement prefixed in code example, and become:

```

namespace RectorPrefix202408;
use RectorPrefix202408\Doctrine\Common\Collections\Criteria;
```

ref https://github.com/rectorphp/rector/commit/b8c4f153f83386ec0da64a2c275564743b4eb438#diff-c260cb91d355f253688b74213c5bb7c75414832d6c18ba756b6ba3accf08964f

Comment (samsonasik) [reply rules/Collection22/Rector/CriteriaOrderingConstantsDeprecationRector.php:35]: https://github.com/rectorphp/rector-doctrine/pull/341

**Comment by TomasVotruba** [comment]
Thank you @julienfastre  :+1:

File: config/sets/doctrine-code-quality.php [hunk 5]
```diff
@@ -31,8 +30,6 @@
         TypedPropertyFromColumnTypeRector::class,
         TypedPropertyFromToOneRelationTypeRector::class,
         TypedPropertyFromToManyRelationTypeRector::class,
-
-        OrderByKeyToClassConstRector::class,
     ]);
```
Comment (TomasVotruba) [inline config/sets/doctrine-code-quality.php:33]: I wonder, why not add the rule here too?

Comment (samsonasik) [reply config/sets/doctrine-code-quality.php:33]: This specifically for doctrine 2.2+

**Pull request merged**